flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaRecordSerializationSchemaBuilder.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.flink.streaming.connectors.kafka.partitioner.FlinkKafkaPartitioner;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.serialization.Serializer;
+
+import javax.annotation.Nullable;
+
+import java.util.Map;
+import java.util.OptionalInt;
+import java.util.function.Function;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Builder to construct {@link KafkaRecordSerializationSchema}.
+ *
+ * <p>This class should give a first entrypoint when trying to serialize elements to {@link
+ * ProducerRecord}. The following examples show some of the possibilities.
+ *
+ * <pre>Simple key-value serialization:
+ * {@code
+ * new KafkaRecordSerializationSchemaBuilder<String>("topic")
+ *     .withKeySerializationSchema(new SimpleStringSchema())
+ *     .withValueSerializationSchema(new SimpleStringSchema())
+ *     .build()
+ * }</pre>
+ *
+ * <pre>Using Kafka's serialization stack:
+ * {@code
+ * new KafkaRecordSerializationSchemaBuilder<String>("topic")
+ *     .withKeySerializer(StringSerializer.class)
+ *     .withValueSerializer(StringSerializer.class)
+ *     .build()
+ * }</pre>
+ *
+ * <pre>With custom partitioner:
+ * {@code
+ * new KafkaRecordSerializationSchemaBuilder<String>("topic")
+ *     .withPartitioner(MY_FLINK_PARTITIONER)
+ *     .withKeySerializationSchema(StringSerializer.class)
+ *     .build()
+ * }</pre>
+ *
+ * <p>The different serialization methods for key and value are mutually exclusive thus i.e. it is
+ * not possible to use {@link #withKeySerializationSchema(SerializationSchema)} and {@link
+ * #withKeySerializer(Class)} on the same builder instance.
+ *
+ * <p>It is necessary to configure exactly one serialization method for the value.
+ *
+ * @param <IN> type of records to be serialized
+ * @see KafkaRecordSerializationSchema#builder(String)
+ */
+public class KafkaRecordSerializationSchemaBuilder<IN> {
+
+    @Nullable private final String topic;
+    @Nullable private final Function<IN, String> topicSelector;
+    @Nullable private FlinkKafkaPartitioner<IN> partitioner;
+    @Nullable private SerializationSchema<IN> keySerializationSchema;
+
+    private SerializationSchema<IN> valueSerializationSchema;
+
+    KafkaRecordSerializationSchemaBuilder(String topic) {
+        this.topic = checkNotNull(topic);
+        this.topicSelector = null;
+    }
+
+    KafkaRecordSerializationSchemaBuilder(Function<IN, String> topicSupplier) {
+        this.topicSelector = checkNotNull(topicSupplier);
+        this.topic = null;
+    }
+
+    /**
+     * Sets a custom partitioner determining the target partition of the target topic.
+     *
+     * @param partitioner
+     * @return {@link KafkaRecordSerializationSchemaBuilder}
+     */
+    public KafkaRecordSerializationSchemaBuilder<IN> withPartitioner(
```

Comment:
Is there also some Kafka partitioner?

Reply:
There is a `FlinkFixedPartitioner` in the Flink Kafka module. By default, Kafka decides how to partition if the partitioner is null.

Reply:
I was thinking about `org.apache.kafka.clients.producer.Partitioner`.

Reply:
So far no one has ever asked to support the native Kafka `Partitioner` I'd prefer to leave it out for now until we find some users.

Reply:
I guess it's easy enough for users to write an adapter if they really need it. So I'm fine with leaving it out.